### PR TITLE
Follow-up fixes on multi-stage moby's Dockerfile

### PR DESCRIPTION
**Pull request by vdemeester**
- Do not copy golang in itself (`runtime-dev` is already base on `golang`)
- Use `golang` official image as base (instead of `debian:stretch` and installing go manually)

This is gonna affect the cache quite a lot (because well, it's updating the base image for almost everything 😅 )

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

**Review by AkihiroSuda — APPROVED** [review]

**Review by boaz0 — APPROVED** [review]
LGTM :clap:

**Comment by ripcurld** [comment]
CI failed. On windows:
```
ERROR: Failed 'You cannot call a method on a null-valued expression.' at 02/27/2018 08:57:28
```
I have no idea why.

In `janky` and `experimental` it's due to flaky tests: `DockerSwarmSuite.TestAPISwarmServicesPlugin` and `DockerSwarmSuite.TestAPISwarmManagerRestore`

File: Dockerfile [hunk 1]
```diff
@@ -32,23 +32,11 @@
 # the case. Therefore, you don't have to disable it anymore.
 #
 
-FROM buildpack-deps:stretch AS base
+FROM golang:1.9.4 AS base
```
Comment (thaJeztah) [inline Dockerfile]: Wondering if we should use:

```Dockerfile
ARG GO_VERSION=1.9.3
FROM golang:$GO_VERSION
```

Which would also allow testing on different versions if we'd want to

File: Dockerfile [hunk 2]
```diff
@@ -32,23 +32,11 @@
...
-
-FROM base AS golang
-# IMPORTANT: If the version of Go is updated, the Windows to Linux CI machines
-#            will need updating, to avoid errors. Ping #docker-maintainers on IRC
-#            with a heads-up.
-# IMPORTANT: When updating this please note that stdlib archive/tar pkg is vendored
-ENV GO_VERSION 1.9.4
-RUN curl -fsSL "https://golang.org/dl/go${GO_VERSION}.linux-amd64.tar.gz" \
-	| tar -xzC /usr/local
-ENV PATH=/usr/local/go/bin:/go/bin:$PATH GOPATH=/go
```
Comment (thaJeztah) [inline Dockerfile]: Guess it doesn't make a difference for us, but the order is slightly different in the official Golang images (just leaving it here);

```Dockerfile
ENV GOPATH /go
ENV PATH $GOPATH/bin:/usr/local/go/bin:$PATH
```

File: Dockerfile (hunk 1 above)
Comment (cpuguy83) [reply Dockerfile]: Actually this means we can't test go rc's

Comment (thaJeztah) [reply Dockerfile]: Should work, Docker Hub also pushes images for rc's

Comment (vdemeester) [reply Dockerfile]: @cpuguy83 to test RC's we could also "temporarly" refer to other images we build too, couldn't we ? 👼

Comment (cpuguy83) [reply Dockerfile]: You can't do anything out of the norm! ;)

Actually didn't realize there were rc's on hub. That's great.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by cpuguy83** [comment]
Except CI failures. :(

**Comment by vdemeester** [comment]
Oh windowsRS1 failure is real... the `windowsRS1` CI script looks for `ENV GO_VERSION`…

```
    $goVersionDockerfile=$(Get-Content ".\Dockerfile" | Select-String "ENV GO_VERSION").ToString().Split(" ")[2]
```

**Comment by dnephin** [comment]
I think you can do ` $goVersionDockerfile=$(go version).ToString().Split(" ")[2]`, but also not super familiar with powershell.

**Review by dnephin — APPROVED** [review]
LGTM

File: Dockerfile (hunk 1 above)
Comment (tianon) [reply Dockerfile]: We have a long tradition of pushing the RCs too: :innocent: 
```console
$ docker-hub-list-tags.sh library/golang | grep rc | grep -vE 'onbuild|wheezy|windowsservercore|nanoserver|cross'
rc
1.10-rc
1.10rc2
rc-stretch
1.10-rc-stretch
1.10rc2-stretch
rc-alpine
1.10-rc-alpine
1.10rc2-alpine
rc-alpine3.7
1.10-rc-alpine3.7
1.10rc2-alpine3.7
1.10rc1
1.10rc1-alpine
1.10rc1-alpine3.7
1.10rc1-stretch
1.9-rc-alpine
1.9rc2-alpine
rc-alpine3.6
1.9-rc-alpine3.6
1.9rc2-alpine3.6
1.9-rc
1.9rc2
1.9-rc-stretch
1.9rc2-stretch
1.9rc1-alpine
1.9rc1-alpine3.6
1.9rc1
1.9rc1-stretch
1.8-rc-alpine
1.8rc3-alpine
1.8-rc-stretch
1.8rc3-stretch
1.8-rc
1.8rc3
1.8rc2-alpine
1.8rc2
1.8rc1
1.8rc1-alpine
1.7rc6-alpine
1.7rc6
1.7rc5-alpine
1.7rc5
1.7rc4-alpine
1.7rc4
1.7rc3
1.7rc3-alpine
1.7rc2-alpine
1.7rc2
1.7rc1-alpine
1.7rc1
1.6rc2-alpine
1.6rc2
1.6rc1-alpine
1.6rc1
1.5rc1
1.4rc2
1.4rc1
```
(back to Go 1.4rc1 -- https://github.com/docker-library/golang/pull/26; 2014! :smile:)

File: Dockerfile [hunk 3]
```diff
@@ -32,23 +32,11 @@
...
 ARG APT_MIRROR=deb.debian.org
 RUN sed -ri "s/(httpredir|deb).debian.org/$APT_MIRROR/g" /etc/apt/sources.list
 
-
-FROM base AS golang
-# IMPORTANT: If the version of Go is updated, the Windows to Linux CI machines
-#            will need updating, to avoid errors. Ping #docker-maintainers on IRC
-#            with a heads-up.
-# IMPORTANT: When updating this please note that stdlib archive/tar pkg is vendored
-ENV GO_VERSION 1.9.4
```
Comment (thaJeztah) [inline Dockerfile]: Actually, just realised that we use the `GO_VERSION` env-var as well in the scripts to generate the "builder" images; https://github.com/moby/moby/blob/4bdb6ed9b58d641176a601cfe92d496038c5a557/contrib/builder/deb/ppc64le/generate.sh#L89

Comment (vdemeester) [reply Dockerfile]: ohhh :sweat: ok I think I have an idea of what to do :angel:

**Review by dnephin — APPROVED** [review]
LGTM

**Review by thaJeztah — APPROVED** [review]
LGTM

**Pull request merged**